[3D] crash - Fix terrain reloading

ptitjano opened the pull request:
## Description

This PR fixes a crash which occurs when a new terrain is loaded while the current terrain is still loading. This can easily be reproduced by quickly clicking 2 times on the `Apply` button of the 3d configuration dialog.

`QgsChunkedEntity` does not properly handle the cancellation and there is no mechanism to stop the `QgsChunkedEntity` jobs from the terrain generator. This PR tries to fix those issues.

From the main commit message: 

When a terrain (a `QgsTerrainEntity`) is replaced by a new one (in
`Qgs3DMapSettings::setTerrainGenerator()`),
`Qgs3DMapScene::createTerrain()` is called. This method removes the
existing `QgsTerrainEntity` and creates a new one. However, if some
jobs are still running on the existing terrain, this results in a
crash because the existing jobs then try to delete objects which have
already been freed by the deletion of the terrain.

This issue is fixed by stopping all the existing jobs and by
preventing the creation of new jobs when a terrain is removed before
the creation of the new one. This is achieved by adding a new public
method `stopAllJobs` to `QgsChunkedEntity`.

cc @wonder-sk @benoitdm-oslandia

ptitjano committed:
qgschunkedentity: Simplify while logic in startJobs()

ptitjano committed:
qgschunkedentity: Disconnect finished signal when a job is canceled

This prevents a crash in `QgsChunkedEntity::onActiveJobFinished()`
which tries to access a job that no long exists.

benoitdm-oslandia commented on src/3d/chunks/qgschunkedentity_p.h:
```diff
@@ -128,6 +128,13 @@ class QgsChunkedEntity : public Qgs3DMapSceneEntity
+    *
+    * \since QGIS 3.34
+    */
+    void stopAllJobs();
```
I am not fond of this name, what do you think about `invalidate`?

wonder-sk commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -472,6 +472,7 @@ void Qgs3DMapScene::createTerrain()
     mSceneEntities.removeOne( mTerrain );
+    mTerrain->stopAllJobs();
 
     mTerrain->deleteLater();
```
what if we just replaced `mTerrain->deleteLater();` by `delete mTerrain;`, without adding stopAllJobs() ? The destructor is expected to cancel all pending jobs. If that would work, that would be a much simpler solution than adding stopAllJobs() and having `mIsValid` as a new state that chunked entity implementation needs to keep in mind...

wonder-sk replied:
cool - if that works, please feel free to remove any extra code we don't need (addition of stopAllJobs()) and we can merge it...

ptitjano replied:
Done.

wonder-sk commented on src/3d/chunks/qgschunkedentity_p.h:
```diff
@@ -128,6 +128,13 @@ class QgsChunkedEntity : public Qgs3DMapSceneEntity
+    /**
+    * Stops all the existing jobs and set the entity as not valid to prevent creating new jobs
+    *
+    * \since QGIS 3.34
```
no need to mark here \since, because it is private API (_p.h)

wonder-sk commented on src/3d/chunks/qgschunkedentity_p.cpp:
```diff
@@ -689,6 +686,11 @@ QgsChunkQueueJob *QgsChunkedEntity::startJob( QgsChunkNode *node )
+    {
+      node->cancelQueuedForLoad();
+      return nullptr;
+    }
```
I am not sure if I understand why it is necessary to add this - my understanding is that when jobs get stopped, no new jobs should get started before the entity gets deleted...

wonder-sk commented on src/3d/chunks/qgschunkedentity_p.cpp:
```diff
@@ -739,6 +748,12 @@ void QgsChunkedEntity::cancelActiveJob( QgsChunkQueueJob *job )
 
+void QgsChunkedEntity::stopAllJobs()
+{
+  mIsValid = false;
```
Do we really need to add `mIsValid` ? It makes the logic more complicated, and code in various places should account for it... It is also not clearly defined what it does - if mIsValid==false, shall we just stop loading nodes? or also stop updating the entity when scene changes?

If we really need an extra flag, I would suggest to have something more explcit - e.g. `mStartJobsAllowed` - that would determine whether startJobs() is allowed to start any new jobs or not. So when the terrain gets deleted, 3D map scene would first call cancelActiveJobs() followed by setStartJobsAllowed(false). This may be useful also in other scenarios - for example when we reach entity's memory limit, we would not start any new jobs until some memory gets freed.

ptitjano replied:
This PR does not introduce the `mIsValid` flag. It was introduced by this [commit](https://github.com/qgis/qGIS/commit/3b86c51b25b311aa06d2509339b54bb6b00e47fb). I'm not exactly sure what it supposed to achieve. It looks like, it prevents generating the terrain if the underlying layer is not valid.

wonder-sk replied:
Ah right, sorry :-) It looks like it is meant to prevent any updates of the scene when the source of the layer is invalid...

ptitjano committed:
qgs3dmapscene: Delete the terrain to properly stop active jobs

When a terrain (a `QgsTerrainEntity`) is replaced by a new one (in
`Qgs3DMapSettings::setTerrainGenerator()`),
`Qgs3DMapScene::createTerrain()` is called. This method removes the
existing `QgsTerrainEntity` by calling `mTerrain->deleteLater()` and
creates a new one. However, if some jobs are still running on the
existing terrain, this results in a crash because the existing jobs
then try to delete objects which have already been freed by the
deletion of the terrain.

This issue is fixed by calling the terrain desctuctor instead. This
calls `cancelActiveJobs` which stops all the existing jobs and
prevents the creation of new jobs. This way, the existing jobs are
correctly stopped.

q#	worktree-geometry-boundingbox/

andreasneumann commented:
Interesting. Maybe it also fixes the quite frequent crashes I often experience with my 3D scenes (usually on opening the 3D view) ?

Hopefully I can try this out soon.

nyalldawson commented:
I'm hopeful this one fixes the long-standing crash you can get if you rapidly change the terrain height -- fingers crossed! 

Great work @ptitjano !

ptitjano commented:
@wonder-sk This PR should be ready now.
